Document CPU limit mode in AMS (#149)
anbox-cloud/discourse-docs: exp/performance.md

Generally, applications should use the smallest suitable instance type. However, if you see an overall bad performance when running the application, using a more powerful instance type usually helps (even though it reduces the container density). As an example, consider an application that runs on an Anbox Cloud deployment that does not have any GPUs installed. In this case, the rendering workload is put on the CPU instead of the GPU, and if the instance type of the application does not have a sufficient number of vCPU cores, the performance of the application is impacted. This can show, for example, in the virtual keyboard being really slow. By switching to a more powerful instance type, the container density is reduced, but the performance of each application container is increased.

<a name="container-cpu-access"></a>
## Container CPU access

AMS has different modes to grant CPU access to a container. The `cpu.limit_mode` configuration option can be used to change the mode. The possible modes are:

> * `scheduler` - This will use the LXD [`limits.cpu.allowance`](https://linuxcontainers.org/lxd/docs/master/instances/#cpu-limits) <!-- wokeignore:rule=master --> configuration option to grant a container a CPU time budget via the Linux CFS scheduler. See [CFS Bandwidth Control](https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html) for more details.

Reviewer: ```suggestion
* `scheduler` - This mode uses the LXD [`limits.cpu.allowance`](https://linuxcontainers.org/lxd/docs/latest/instances/#cpu-limits) configuration option to grant a container a CPU time budget via the Linux CFS scheduler. See [CFS Bandwidth Control](https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html) for more details.
```
We have "latest" now as an alternative :)

Author: Ah good to know :-)

> * `pinning` - This will use the LXD [`limits.cpu`](https://linuxcontainers.org/lxd/docs/master/instances/#cpu-limits) <!-- wokeignore:rule=master --> configuration option to pin a set of CPU cores to a container. LXD is responsible to allocate a specific number of cores to a container and load balance all running containers on all available cores.

Author: The container gets pinned to specific CPUs but which ones LXD is going to decide. If you look at the cpuset configuration of the container you will see specific cores being listed. LXD will load balance container across all available cores to find the best distribution.

> * `pinning` - This will use the LXD [`limits.cpu`](https://linuxcontainers.org/lxd/docs/master/instances/#cpu-limits) <!-- wokeignore:rule=master --> configuration option to pin a set of CPU cores to a container. LXD is responsible to allocate a specific number of cores to a container and load balance all running containers on all available cores.

Reviewer 2: Yes, the explanation makes more sense. The term pinning used here introduces a little bit confusing to me. Since a **different** set of CPU cores will be allocated by LXD dynamically and pinned to **different** containers. So each CPU core may serve one or N containers depending on the number of CPU cores a host has,  also the number of containers run simentensoly.
I am fine to leave it as is.

> * `pinning` - This will use the LXD [`limits.cpu`](https://linuxcontainers.org/lxd/docs/master/instances/#cpu-limits) <!-- wokeignore:rule=master --> configuration option to pin a set of CPU cores to a container. LXD is responsible to allocate a specific number of cores to a container and load balance all running containers on all available cores.

Reviewer: ```suggestion
* `pinning` - This mode uses the LXD [`limits.cpu`](https://linuxcontainers.org/lxd/docs/latest/instances/#cpu-limits) configuration option to pin a set of CPU cores to a container. LXD is responsible for allocating a specific number of cores to a container and load-balance all running containers on all available cores.
```

> By default AMS uses the `scheduler` option as it provides the most generic solution to a large set of use cases Anbox Cloud supports. However in some cases CPU pinning might be the better option to distribute load across all available CPU cores on a system.

Reviewer: ```suggestion
By default, AMS uses the `scheduler` option, because it provides the most generic solution to a large set of use cases  that Anbox Cloud supports. However, in some cases CPU pinning might be the better option to distribute load across all available CPU cores on a system.
```

Reviewer: Can we give an example when it might be the better option?

Reviewer: Do we have an example?

Author: Not a good one as it really depends on various individual aspects.

<a name="container-cpu-access"></a>
## Container CPU access

AMS has different modes to grant CPU access to a container. The `cpu.limit_mode` configuration option can be used to change the mode. The possible modes are:

* `scheduler` - This mode uses the LXD [`limits.cpu.allowance`](https://linuxcontainers.org/lxd/docs/latest/instances/#cpu-limits) configuration option to grant a container a CPU time budget via the Linux CFS scheduler. See [CFS Bandwidth Control](https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html) for more details.
* `pinning` - This mode uses the LXD [`limits.cpu`](https://linuxcontainers.org/lxd/docs/latest/instances/#cpu-limits) configuration option to pin a set of CPU cores to a container. LXD is responsible for allocating a specific number of cores to a container and load-balance all running containers on all available cores.

By default, AMS uses the `scheduler` option, because it provides the most generic solution to a large set of use cases  that Anbox Cloud supports. However, in some cases CPU pinning might be the better option to distribute load across all available CPU cores on a system.

> [note type="information" status="Note"]Using `pinning` will require a system with [cgroup-v2](https://docs.kernel.org/admin-guide/cgroup-v2.html) enabled, as otherwise limitations of [cgroup-v1](https://docs.kernel.org/admin-guide/cgroup-v1/index.html) will cause not optimal load distribution over available CPU cores. [cgroup-v2](https://docs.kernel.org/admin-guide/cgroup-v2.html) is enabled by default starting with Ubuntu 22.04 and can be enabled on Ubuntu 20.04 by booting with `systemd.unified_cgroup_hierarchy=1` added to [the kernel boot parameters](https://wiki.ubuntu.com/Kernel/KernelBootParameters).[/note]

Author: @ru-fu I've added this paragraph to document the cgroup-v2 requirement we have for `pinning`

> [note type="information" status="Note"]Using `pinning` will require a system with [cgroup-v2](https://docs.kernel.org/admin-guide/cgroup-v2.html) enabled, as otherwise limitations of [cgroup-v1](https://docs.kernel.org/admin-guide/cgroup-v1/index.html) will cause not optimal load distribution over available CPU cores. [cgroup-v2](https://docs.kernel.org/admin-guide/cgroup-v2.html) is enabled by default starting with Ubuntu 22.04 and can be enabled on Ubuntu 20.04 by booting with `systemd.unified_cgroup_hierarchy=1` added to [the kernel boot parameters](https://wiki.ubuntu.com/Kernel/KernelBootParameters).[/note]

Reviewer: ```suggestion
[note type="information" status="Note"]
Using `pinning` requires a system with [cgroup-v2](https://docs.kernel.org/admin-guide/cgroup-v2.html) enabled. Otherwise, limitations of [cgroup-v1](https://docs.kernel.org/admin-guide/cgroup-v1/index.html) might cause the load distribution over available CPU cores to not be optimal. [cgroup-v2](https://docs.kernel.org/admin-guide/cgroup-v2.html) is enabled by default starting with Ubuntu 22.04 and can be enabled on Ubuntu 20.04 by booting with `systemd.unified_cgroup_hierarchy=1` added to [the kernel boot parameters](https://wiki.ubuntu.com/Kernel/KernelBootParameters).
[/note]
```
